File: scripts/compat/shims/driver.go
// Exercises clients/go/ttid.go against the binary in TTID_BIN.
package main

import (
	"encoding/json"
	"fmt"
	"os"

	"ttidshim/ttid"
)

const (
	fixed   = "4SQ1NZT5HC0"
	updated = "4SQ1NZT5HC0-4SQ1NZT5P1S"
	deleted = "4SQ1NZT5HC0-4SQ1NZT5P1S-4SQ1NZT5WRK"
)

type row struct {
	name  string
	value any
}

func main() {
	t, err := ttid.Open(os.Getenv("TTID_BIN"))
	if err != nil {
		panic(err)
	}
	defer t.Close()

	must := func(v any, err error) any {
		if err != nil {
			panic(err)
		}
		return v
	}

	out := []row{
		{"generate", must(t.Generate("", false))},
		{"update", must(t.Generate(fixed, false))},
		{"delete", must(t.Generate(updated, true))},
		{"decode", must(t.DecodeTime(deleted))},
		{"isTTID", must(t.IsTTID(fixed))},
		{"isTTID-bad", must(t.IsTTID("nope"))},
		{"isUUID", must(t.IsUUID("3f2504e0-4f89-41d3-9a0c-0305e82c3301"))},
		{"isUUID-bad", must(t.IsUUID("nope"))},
	}
	if _, err := t.Generate(deleted, false); err != nil {
		out = append(out, row{"error", err.Error()})
	} else {
		out = append(out, row{"error", "NO ERROR RAISED"})
	}
	for _, r := range out {
		// encoding/json sorts map keys already.
		b, _ := json.Marshal(r.value)
		fmt.Printf("%s=%s\n", r.name, b)
	}
}

File: scripts/compat/shims/Driver.java
// Exercises clients/java/Ttid.java against the binary in TTID_BIN.
// The client returns raw response lines; the harness masks durationMs.
public final class Driver {
    private static final String FIXED = "4SQ1NZT5HC0";
    private static final String UPDATED = "4SQ1NZT5HC0-4SQ1NZT5P1S";
    private static final String DELETED = "4SQ1NZT5HC0-4SQ1NZT5P1S-4SQ1NZT5WRK";

    public static void main(String[] args) throws Exception {
        try (Ttid t = new Ttid(System.getenv("TTID_BIN"))) {
            System.out.println("generate=" + t.generate());
            System.out.println("update=" + t.generate(FIXED));
            System.out.println("delete=" + t.generate(UPDATED, true));
            System.out.println("decode=" + t.decodeTime(DELETED));
            System.out.println("isTTID=" + t.isTTID(FIXED));
            System.out.println("isTTID-bad=" + t.isTTID("nope"));
            System.out.println("isUUID=" + t.isUUID("3f2504e0-4f89-41d3-9a0c-0305e82c3301"));
            System.out.println("isUUID-bad=" + t.isUUID("nope"));
            try {
                t.generate(DELETED);
                System.out.println("error=NO ERROR RAISED");
            } catch (Exception e) {
                System.out.println("error=" + e.getMessage());
            }
        }
    }
}
